README
SegFilt version 1.0
=======================

NAME:

SEG Filter


DESCRIPTION:


This is a c++ implementation of SEG program written by Wootton and Federhen made for identifying and masking segments of low
complexity in amino acid sequences. 



SYNOPSIS:
    
    #include<string>
    #include<SegFilt.hpp>

    string ProtSeq = "VGRGVQIGSPHHHHHHHHHHPQPATYQTSGNLGVSYSHSSCGPSYGSQNFSAPYSPYALNQEADVSGGYPQCAPAVYSGNLSSPMVQHHHHHQGYAGGAVGSPQYIHHSYGQEHQSLALATYN";

/* Make object */

    /* Construction */
    SEG<int> seg;
    /*  OR  */
    SEG<int> seg(arg); // arg is : unordered_map<string, int|long|unsigned|double>


/*  Functions */


    string mask = seg.SegFilt(ProtSeq);   
    //mask =  VGRGVQIGSPxxxxxxxxxxPQPATYQTSGNLGVSYSHSSCGPSYGSQNFSAPYSPYALNQEADVSGGYPQCAPAVYSGNLSSPMVxxxxxxxGYAGGAVGSPQYIHHSYGQEHQSLALATYN







CONSTRUCTION AND RUNTIME COMPLEXITY:


     -


ACKNOWLEDGMENTS:

     
Wootton, J. C. and S. Federhen (1993).  Statistics of local complexity in amino
acid sequences and sequence databases.  Computers and Chemistry 17:149-163.

Vakatov D, editor. The NCBI C++ Toolkit Book. Bethesda (MD): National Center for 
Biotechnology Information (US); 2004-. 
Available from: http://www.ncbi.nlm.nih.gov/toolkit/doc/book/ch_getcode_svn


AUTHOR:

Robert Bakaric <rbakaric@irb.hr>, <bakaric@evolbio.mpg.de>

COPYRIGHT AND LICENSE:

 * Copyright 2015 Robert Bakaric
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.



ChangeLog
2015-04-13  Robert Bakaric  <rbakaric@irb.hr>
	* Created! 


src/include/SEG.hpp
/*
 * SGE.hpp
 * 
 * Copyright 2015 Robert Bakaric <rbakaric@irb.hr>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */
 
 
 /* NOTE:
  * 
  * This library has been written according to seg(.c) program:
  * 
  *    Wootton, J.C., Federhen, S. (1993)  Statistics of local complexity
  *    in amino acid sequences and sequence  databases.  Computers &
  *    Chemistry 17: 149-163.
  * 
  * In order to preserve the logic, underlying  data structures and code snippetss  
  * used here are based on the exact code fragments form seg.c (and associated) 
  * source code. Therefore  labeles are marked using same or similar variable names. 
  * 
  * */
 
 
#include <vector>
#include <string>
#include <cstdio>
#include <cstdlib>
#include <algorithm>
#include <math.h>
#include <unordered_map>
#include <lnFact.hpp>


using namespace std; 


/*FIXME: Remove legacy code !!*/

/*FIXME: This needs to be fragmented. 
 * Modularity above all !!*/

template <typename Tint>
class SEG {

/* Globals */
  Tint    SegWindow;
  double SegLocut;
  double SegHicut;
  Tint    MaxX;
  Tint    MaxTrim; 
  Tint    Period;
  Tint    MergeOverlaps;


/* Structures */
  struct Alphabet{  
    Tint   alphasize;           /* size */
    double lnalphasize;         /* ln(size) */
    Tint*  alphaindex;          /* ASCII code table*/ 
    unsigned char* alphaflag;   /* array of indicators if the characer is (not) AA */
  } ;

  struct CSeq{  
    struct CSeq* parent;        /* current one */
    char*  seq;                 /* AA sequence */
    Alphabet* palpha;           /* alphabet info */
    Tint   start;               /* starting for seg. */
    Tint   length;              /* sequence length */
    Tint   Xes;                 /* the number of X's */  
    Tint*  charfreq;            /* number of characters in a string */
    Tint*  state;                
    double entropy;
  } ;

  struct SeqSeg{ 
    Tint begin;          
    Tint end;            
    struct SeqSeg *next;  
  } ;
  

  Alphabet* alpha;
  
  

/* Functions */
   
  void     SegFree(SeqSeg* seg);
  void     MergeSegs(CSeq* seq, SeqSeg* segs);
  template<typename Targ>
  void     SetParamaters(Targ &arg);
  CSeq*    NewCSeq();
  double*  ComputeEntropy(CSeq* seq,Tint  first, Tint last, Tint downset, Tint upset);
  Tint     SegSeq(CSeq* seq, SeqSeg **segs, Tint offset);
  Tint     LocLow(Tint i, Tint limit, double* H);
  Tint     LocHigh(Tint i, Tint limit, double* H);
  CSeq*    OpenWin(CSeq* parent, Tint start, Tint length);
  void     StateOn(CSeq* win);
  void     CompOn(CSeq* win);
  Tint     Trim(CSeq* seq, Tint* leftend, Tint* rightend);
  void     CloseWin(CSeq* win);
  bool     ShiftWin1(CSeq* win);
  double   Entropy(Tint* sv);
  void     DecrementSV(Tint* sv, Tint clas);
  void     IncrementSV(Tint* sv, Tint clas);
  double   GetProb(const Tint* sv, Tint total, const Alphabet* palpha);
  double   LnPerm(const Tint* sv, Tint window_length);
  double   LnAss(const Tint* sv, Tint alphasize);
  double   lnFact(Tint n) ;
  void     CSeqFree(CSeq* seq);
  Alphabet* MakeAlpha();
  void     AlphaFree (Alphabet* alpha);
  template <typename T>
  void     SafeFree(T *x);
  template <typename T>
  void     SafeFree(T **x);
  void     EntropyOn(CSeq* win);
  
   
  public:

/* Constructors */
  template <typename Targ>
    SEG(Targ& arg);
    SEG();

/* Destructors */
  ~SEG();

/* Functions */
   string SegFilt(string str);

};
 


/* Constructors */
template <typename Tint>
SEG<Tint>::SEG(){
  unordered_map<string,Tint> para;
  SetParamaters(para);
  alpha= MakeAlpha();
}; 

template <typename Tint>
template <typename Targ>
SEG<Tint>::SEG(Targ& arg){
  SetParamaters(arg);
  alpha= MakeAlpha();
}
/* Explicite missing*/



/* Destructors */
template <typename Tint>
SEG<Tint>::~SEG(){
    AlphaFree(alpha);
}
/*Explicite missing*/


/* Functions  : Public */

template <typename Tint>
string SEG<Tint>::SegFilt(string str){

  
  CSeq* seq;
  SeqSeg* segs;
  bool params_allocated = false;
  Tint status = 0;
  SeqSeg  *seg;
  Tint begin, end, i;

/* old schoole - parse */

  seq = NewCSeq();
  seq->seq = &str[0];
  seq->length = str.size();
  seq->palpha = alpha;
  
  segs = (SeqSeg*) NULL;

/* compute lc segments */
   status = SegSeq (seq, &segs, 0);
   if (status < 0){
     seq->seq = NULL;
     CSeqFree (seq);
     throw runtime_error ("Low complexity segment computation could not be preformed!" ); 
   }
   
/* Create filtered sequence*/
    string filtstr(seq->seq);

/* merge segment if specified here you can completly omitt
 * this if raw positions are required by def is set to 1
 * - further testing required - 
 */
  if (MergeOverlaps == 1){
      MergeSegs(seq, segs);
      
/* Mask: X it*/
    for (seg=segs; seg!=NULL; seg=seg->next) {
      begin = seg->begin;
      end = seg->end;
      memset(&filtstr[0] + begin, 'x', end - begin +1);
    }
}

/* clean up */
   seq->seq = NULL;
   CSeqFree(seq);
   SegFree(segs);
   
   return filtstr;	
}


/* Functions  : Private */


template <typename Tint>
template <typename Targ>
void SEG<Tint>::SetParamaters(Targ &arg){
  
  (arg.find("window") == arg.end() || arg["window"] <= 0)     ? ( SegWindow = 12 ) : ( SegWindow = arg["window"] );
  (arg.find("hicut") == arg.end() ||  arg["hicut"] < 0)      ? ( SegHicut = 2.5 ) : ( SegHicut  = arg["hicut"] );
  (arg.find("locut") == arg.end() || arg["locut"] < 0)       ? ( SegLocut = 2.2 ) : ( SegLocut  = arg["locut"] );
  (arg.find("maxXes") == arg.end() || arg["maxXes"] < 0) ? ( MaxX = 0)   : ( MaxX  = arg["maxXes"]);
  (arg.find("maxtrim") == arg.end() || arg["maxtrim"] < 0)   ? ( MaxTrim = 100 )    : ( MaxTrim   = arg["maxtrim"]);
  (arg.find("period") == arg.end() || arg["period"] < 1)     ? ( Period = 1 )     : ( Period    = arg["period"]);
  (arg.find("merge") == arg.end() || arg["merge"] < 1)     ? ( MergeOverlaps = 1 )     : ( MergeOverlaps    = arg["merge"]);
  
  
   if ( SegLocut > SegHicut)
       SegHicut = SegLocut;

   if (MaxX > SegWindow)
       MaxX = SegWindow;
}

template <typename Tint>
void SEG<Tint>::SegFree(SeqSeg* seg){
   SeqSeg* nextseg;
   while (seg){
      nextseg = seg->next;
      SafeFree(seg);
      seg = nextseg;
   }
}

template <typename Tint>
 void SEG<Tint>::MergeSegs(CSeq* seq, SeqSeg* segs){
	 
   SeqSeg* seg,* nextseg;          

   if (segs==NULL) return;

   if ((seq->length - 1 - segs->end) < 0) 
       segs->end = seq->length -1;

   seg = segs;
   nextseg = seg->next;

   while (nextseg!=NULL) {
      if (seg->begin - nextseg->end - 1 < 0) {
         if ((seg->end) < (nextseg->end)) seg->end = nextseg->end;
         if ((seg->begin) > (nextseg->begin)) seg->begin = nextseg->begin;
         seg->next = nextseg->next;
         SafeFree(nextseg);
      } else {
         seg = nextseg;
      }
      nextseg = seg->next;
   }
   if ((seg->begin) < 0) seg->begin = 0;
}

template <typename Tint>
typename SEG<Tint>::CSeq*  SEG<Tint>::NewCSeq(){
   CSeq* seq;

   seq = (CSeq*) calloc(1, sizeof(CSeq));
   if (seq==NULL)
      return(seq);

   seq->parent = (CSeq*) NULL;
   seq->seq = (char*) NULL;
   seq->palpha = (Alphabet*) NULL;
   seq->start = 0;
   seq->length = 0;
   seq->Xes =false;
   seq->charfreq = seq->state = (Tint*) NULL;
   seq->entropy = (double) 0.0;

   return(seq);
}

template <typename Tint>
double* SEG<Tint>::ComputeEntropy(CSeq* seq,Tint  first, Tint last, Tint downset, Tint upset){
	
   CSeq* win;
   double* H;
   Tint i;

   if (SegWindow>seq->length)
      return((double*) NULL);
     

   H = (double*) calloc(seq->length, sizeof(double));

   for (i=0; i<seq->length; i++)
      H[i] = -1.0;

   win = OpenWin(seq, 0, SegWindow);
   EntropyOn(win);

   for (i=first; i<=last; i++){
      if ((win->Xes )> MaxX){
         H[i] = -1.;
         ShiftWin1(win);
         continue;
        }
      H[i] = win->entropy;
      ShiftWin1(win);
     }

   CloseWin(win);
   return(H);
}

template <typename Tint>
Tint SEG<Tint>::SegSeq(CSeq* seq, SeqSeg **segs, Tint offset){
   SeqSeg* seg = (SeqSeg*) NULL;

   Tint downset, upset;
   Tint first, last, lowlim;
   Tint i;
   Tint leftend, rightend;
   double* H;
   Tint status = 0;

   downset = (SegWindow+1)/2 - 1;
   upset = SegWindow - downset;
   first = downset;
   last = seq->length - upset;
   lowlim = first;
   
   H = ComputeEntropy(seq,  first,  last,  downset,  upset);

   if (H == NULL) 
      return status;
   for (i=first; i<=last; i++){
      if (H[i] <= SegLocut && H[i] != -1.0){
         Tint loi = LocLow(i, lowlim, H); 
         Tint hii = LocHigh(i, last, H);
         CSeq* temp_seq = NULL;

         leftend = loi - downset;
         rightend = hii + upset - 1;

         temp_seq = OpenWin(seq, leftend, rightend-leftend+1);
         status = Trim(temp_seq, &leftend, &rightend); 

         if (status < 0) {
             CloseWin(temp_seq);
             break;
         }

         if (i+upset-1<leftend){
            Tint lend = loi - downset;
            Tint rend = leftend - 1;

            CSeq* leftseq = OpenWin(seq, lend, rend-lend+1);
            SeqSeg *leftsegs = (SeqSeg*) NULL;
            status = SegSeq(leftseq,  &leftsegs, offset+lend);
            if (status < 0)
              return status;

            if (leftsegs!=NULL){
               leftsegs->next = *segs;
               *segs = leftsegs;
            }
            CloseWin(leftseq);
         }

         seg = (SeqSeg*) calloc(1, sizeof(SeqSeg));
         seg->begin = leftend + offset;
         seg->end = rightend + offset;
         seg->next = *segs;
         *segs = seg;
         i = min(hii, rightend+downset);
         lowlim = i + 1;
        }
   }
   SafeFree(H);
   return status;
}

template <typename Tint>
Tint SEG<Tint>::LocLow(Tint i, Tint limit, double* H){
   Tint j;
   for (j=i; j>=limit; j--){
      if (H[j]==-1.0) break;
      if (H[j]>SegHicut) break;
   }
   return(j+1);
}

template <typename Tint>
Tint SEG<Tint>::LocHigh(Tint i, Tint limit, double* H){
   Tint j;
   for (j=i; j<=limit; j++){
      if (H[j]==-1.0) break;
      if (H[j]>SegHicut) break;
   }
   return(j-1);
}

template <typename Tint>
typename SEG<Tint>::CSeq* SEG<Tint>::OpenWin(CSeq* parent, Tint start, Tint length){
   CSeq* win;

   if (start<0 || length<0 || start+length>parent->length)
      return((CSeq*) NULL);

    win = (CSeq*) calloc(1, sizeof(CSeq));
    win->parent = parent;
    win->palpha = parent->palpha;
    win->start = start;
    win->length = length;
    win->seq = parent->seq + start;
    win->Xes = 0;
    win->entropy = -2.;
    win->state = (Tint*) NULL;
    win->charfreq = (Tint*) NULL;
	
    StateOn(win);

    return win;
}

template <typename Tint>
void SEG<Tint>::StateOn(CSeq* win){
	Tint letter, nel, c;
    Tint alphasize =  win->palpha->alphasize;

	if (win->charfreq == NULL)
		CompOn(win);
	win->state = (Tint*) calloc((alphasize+1), sizeof(win->state[0]));
	for (letter = nel = 0; letter < alphasize; ++letter) {
		if ((c = win->charfreq[letter]) == 0)
			continue;
		win->state[nel++] = c;
	}
	for (letter = nel; letter < alphasize+1; ++letter)
		win->state[letter] = 0;

    sort(win->state,  win->state+nel,greater<Tint>());
}

template <typename Tint>
void SEG<Tint>::CompOn(CSeq* win){
  Tint* comp;
  Tint letter;
  char* seq = win->seq; 
  char* seqmax = seq + win->length;
  Tint* alphaindex = win->palpha->alphaindex;
  unsigned char* alphaflag = win->palpha->alphaflag;
  Tint alphasize = win->palpha->alphasize;

  win->charfreq = comp = (Tint*) calloc(alphasize, sizeof(Tint));

  while (seq < seqmax) {
    letter = *seq++;
    if (!alphaflag[letter])
      comp[alphaindex[letter]]++;
    else 
      win->Xes++;
  }
}

template <typename Tint>
Tint SEG<Tint>::Trim(CSeq* seq, Tint* leftend, Tint* rightend){
   double prob, minprob = 1;
   Tint len;
   Tint lend=0 ;
   Tint rend =seq->length - 1;
   Tint minlen = 1;
   Tint status = 0;

   if ((seq->length-MaxTrim)>minlen) 
        minlen = seq->length-MaxTrim;
  
   for (len=seq->length; len>minlen; len--){
      bool shift = true;
      Tint i = 0;
      CSeq* win = OpenWin(seq, 0, len);

      while (shift){
         prob = GetProb(win->state, len, win->palpha);
         if (prob<minprob)
         {
            minprob = prob;
            lend = i;
            rend = len + i - 1;
         }
         shift = ShiftWin1(win);
         i++;
      }
      CloseWin(win);
   }

   *leftend = *leftend + lend;
   *rightend = *rightend - (seq->length - rend - 1);

   CloseWin(seq);
   return status;
}

template <typename Tint>
void SEG<Tint>::CloseWin(CSeq* win)
{
   if (win==NULL) return;

   if (win->state!=NULL)       SafeFree(win->state);
   if (win->charfreq!=NULL) SafeFree(win->charfreq);

   SafeFree(win);
   return;
}

template <typename Tint>
bool SEG<Tint>::ShiftWin1(CSeq* win){
	
  Tint j, length = win->length;
  Tint* comp = win->charfreq;
  Tint* alphaindex = win->palpha->alphaindex;
  unsigned char* alphaflag = win->palpha->alphaflag;

  if ((++win->start + length) > win->parent->length) {
    --win->start;
    return false;
  }

  if (!alphaflag[j = win->seq[0]])
    DecrementSV(win->state, comp[alphaindex[j]]--);
  else 
    win->Xes--;
  j = win->seq[length];
  ++win->seq;

  if (!alphaflag[j])
    IncrementSV(win->state, comp[alphaindex[j]]++);
  else 
    win->Xes++;

  if (win->entropy > -2.)
    win->entropy = Entropy(win->state);

	return true;
}

template <typename Tint>
double SEG<Tint>::Entropy(Tint* sv){
   double ent;
   Tint i, total = 0;

   for (i=0; sv[i]!=0; i++)
     {
      total += sv[i];
     }
   if (total==0) return(0.);

   ent = 0.0;
   const double ln2 = 0.69314718055994530941723212145818;
   for (i=0; sv[i]!=0; i++)
      ent += ((double)sv[i])*log(((double)sv[i])/(double)total)/ln2;   /// WHAT IS THIS
   double f = ent/(double)total; 
   ent = (f < 0.0) ? -(f) : (f);

   return(ent);
}

template <typename Tint>
void SEG<Tint>::DecrementSV(Tint* sv, Tint clas){
  Tint	svi;
  while ((svi = *sv++) != 0) {
    if (svi == clas && *sv < clas) {
      sv[-1] = svi - 1;
    break;
    }
  }
}

template <typename Tint>
void SEG<Tint>::IncrementSV(Tint* sv, Tint clas){
  for (;;) {
    if (*sv++ == clas) {
      sv[-1]++;
      break;
    }
  }
}

template <typename Tint>
double SEG<Tint>::GetProb(const Tint* sv, Tint total, const Alphabet* palpha){
   double  ans1, ans2 = 0, totseq;

   totseq = ((double) total) * (palpha->lnalphasize);

   ans1 = LnAss(sv, palpha->alphasize);
   if (ans1 > -100000.0 && sv[0] != (-2147483647-1))  // ncbi constant
      ans2 = LnPerm(sv, total);
   else
      cerr <<"Bad value!\n";
   
   return  ans1 + ans2 - totseq;
}
  
template <typename Tint>
double SEG<Tint>::LnPerm(const Tint* sv, Tint window_length){
   double ans;
   Tint i;
   
   ans = lnFact(window_length);
   for (i=0; sv[i]!=0; i++)
      ans -= lnFact(sv[i]);

   return(ans);
}

template <typename Tint>
double SEG<Tint>::LnAss(const Tint* sv, Tint alphasize){
  double	ans;
  Tint	svi, svim1;
  Tint	clas, total;
  Tint    i;

  ans = lnfact[alphasize];
  if (sv[0] == 0)
    return ans;

  total = alphasize;
  clas = 1;
  svi = *sv;
  svim1 = sv[0];
  for (i=0;; svim1 = svi) {
    if (++i==alphasize) {
      ans -= lnFact(clas);
      break;
    }else if ((svi = *++sv) == svim1) {
      clas++;
      continue;
    }else {
      total -= clas;
      ans -= lnFact(clas);
      if (svi == 0) {
        ans -= lnFact(total);
        break;
      }else {
        clas = 1;
        continue;
      }
    }
  }
  return ans;
}
 
template <typename Tint>
double SEG<Tint>::lnFact(Tint n) {
  if (n < sizeof(lnfact)/sizeof(*lnfact))
     return lnfact[n];
  else 
    return ((n+0.5)*log(n) - n + 0.9189385332);
}

template <typename Tint>
void SEG<Tint>::CSeqFree(CSeq* seq){
   if (seq==NULL) return;
   SafeFree(seq->seq);
   SafeFree(seq->charfreq);
   SafeFree(seq->state);
   SafeFree(seq);
}

template <typename Tint>
typename SEG<Tint>::Alphabet* SEG<Tint>::MakeAlpha (){
   Alphabet* palpha;
   Tint* alphaindex;
   unsigned char* alphaflag;
   Tint c,  i;
   const int kCharSet = 128;
   const double kLn20 = 2.9957322735539909;  // ncbi 

   palpha = (Alphabet*) calloc(1, sizeof(Alphabet));

   palpha->alphasize = 20;
   palpha->lnalphasize = kLn20;

   alphaindex = (Tint*) calloc(kCharSet , sizeof(Tint));
   alphaflag = (unsigned char*) calloc(kCharSet , sizeof(char));

   for (c=0, i=0; c<kCharSet; c++)
     {
        if (c == 65 || 
           (c >= 67 && c <= 73) || 
           (c >= 75 && c <= 78) || 
           (c >= 80 && c <= 84) || 
           (c >= 86 && c <= 87) ||  
           c == 89) {
			   
           alphaflag[c] = false; 
           alphaflag[c+32] = false;
           alphaindex[c] = i; 
           alphaindex[c+32] = i;
           ++i;
        } else {
           alphaflag[c] = true; 
           alphaindex[c] = 20;
        }
     }

   palpha->alphaindex = alphaindex;
   palpha->alphaflag = alphaflag;

   return (palpha);
}

template <typename Tint>
void SEG<Tint>::AlphaFree (Alphabet* alpha){
   SafeFree (alpha->alphaindex);
   SafeFree (alpha->alphaflag);
   SafeFree (alpha);
}

template <typename Tint>
void  SEG<Tint>::EntropyOn(CSeq* win){
   if (win->state==NULL) 
     StateOn(win);
   win->entropy = Entropy(win->state);
}

template <typename Tint>
template <typename T>
void SEG<Tint>::SafeFree(T **x){
    free(*x);
    *x = NULL;

}

template <typename Tint>
template <typename T>
void SEG<Tint>::SafeFree(T *x){
    free(x);
    x = NULL;

}


src/apps/SegFilt.cpp
/*
 * SegFilt.cpp
 * 
 * Copyright 2015 Robert Bakaric <rbakaric@irb.hr>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */




#include<iostream>
#include<fstream>
#include<string>
#include<unordered_map>
#include<boost/program_options.hpp>

#include<SEG.hpp>


namespace po = boost::program_options;
using namespace std;



void PrintAcknowledgements(){

cout <<"***************************************************************************************"<<"\n";
cout <<"                                 SegFilt - SEG Filter                                  "<<"\n";
cout <<"                                          by                                           "<<"\n";
cout <<"                                    Robert Bakaric                                     "<<"\n\n";
cout <<"CONTACT:                                                                               "<<"\n";
cout <<" Code written and maintained by Robert Bakaric,                                        "<<"\n";
cout <<" email: rbakaric@irb.hr , bakaric@evolbio.mpg.de                                       "<<"\n\n";
cout <<"ACKNOWLEDGEMENT:                                                                       "<<"\n";
cout <<"   Wootton, J. C. and S. Federhen (1993).  Statistics of local complexity in amino     "<<"\n";
cout <<"   acid sequences and sequence databases.  Computers and Chemistry 17:149-163.         "<<"\n";
cout <<"                                                                                       "<<"\n";
cout <<"LICENSE:                                                                               "<<"\n";
cout <<" The program is distributed under the GNU General Public License. You should have      "<<"\n";
cout <<" received a copy of the licence together  with this software. If not, see              "<<"\n";
cout <<" http://www.gnu.org/licenses/                                                          "<<"\n";
cout <<"***************************************************************************************"<<"\n\n\n";
}


template <typename INT, typename CHARA>
po::variables_map SetOptions(INT& argc, CHARA& argv){

    try {
        int opt;
        string version = "1.0";
        po::options_description desc("Allowed options");
        desc.add_options()
            ("help,h", "produce help message")
            ("version,v", "print version information")
            ("input-file,i", po::value< string >(), "Fasta input file")
            ("window,w", po::value< int >()->default_value(12), "SEG window size")
            ("locut,L", po::value< double >()->default_value(2.2), "Low complexity cutoff (starter)")
            ("hicut,H", po::value< double >()->default_value(2.5), "High complexity cutoff (starter)")
            ("maxXes,x", po::value< int >()->default_value(0), "Maximum number of xxx  symboles tolerated (dynamically defined if left unchanged)")
            ("maxTrim,t", po::value< int >()->default_value(100),"Maximum trimming of raw segment")
        ;

        po::positional_options_description p;
        p.add("input-file,i", -1);
        
        po::variables_map vm;
        po::store(po::command_line_parser(argc, argv).
                  options(desc).positional(p).run(), vm);
        po::notify(vm);
    
        if (vm.count("help")) {
            cout << "Usage: ./program [options]\n\n";
            PrintAcknowledgements();
            cout << desc;
            exit(0);
        }else if (vm.count("version")) {
            cout << "Program version:  " << version << "\n";
            exit(0);
        }
        if (!vm.count("input-file")){
            cout << "Input file is not defined \n";
            exit(0);
        }
        return vm;    
    }
    catch(std::exception& e)
    {
        cout << e.what() << "\n";
        exit(0);
    }    
}

template <typename Tseg>
void FilterFile(const string file, Tseg& seg){
  
  fstream fs;
  
  fs.open (file.c_str(), ios::in);
  
  if ( !fs.is_open())
    throw  runtime_error ("Cannot open file: " + file );

  string str;
  char c;
  int i=0, f=0;
  
  while( fs.good()) {
    c = fs.get();
    if(fs.eof())
       break;
    if (c == '>'){
		if (str.size()> 0){
            str.resize(i);
            string s = seg.SegFilt(str);
            cout << s << endl;
            str.resize(0);
            i=0;
		}
        f=1;
        cout<<c;
     }else if( f == 1 && c == '\n'){
        f=0;
        cout<<c;
     }else if( f == 1 ){
        cout<<c;
     }else if (f == 0 && c !='\n'){
		 if(i == str.size() )
		     str.resize((str.size()+1)*2);
		 str[i++] = c;
	 }
  }

/* Flush */
str.resize(i);
string s = seg.SegFilt(str);
cout << s << endl;
fs.close();
}



int main (int argc, char** argv){

/* Get cmd Options */  
  po::variables_map arg;
  arg = SetOptions(argc,argv);
  

/* Parse Options */
  unordered_map<string, int> args;
  
  args["window"] = static_cast<int>(arg["window"].as<int>());
  args["hicut"] =  static_cast<double>(arg["hicut"].as<double>());
  args["locut"] =  static_cast<double>(arg["locut"].as<double>());
  args["maxtrim"] =  static_cast<int>(arg["maxTrim"].as<int>());
  args["maxXes"] =  static_cast<int>(arg["maxXes"].as<int>());

  try{

/* Make object */
    SEG<int> seg(args);
  
/* Filter Sequences */  
    FilterFile(arg["input-file"].as<string>(), seg);
    
  }catch(runtime_error& e){
    cerr << e.what() << "\n";
  }
  
  return 0;
}
